JAMES-3138 recompute quotas task & endpoint

Arsnael opened the pull request:
Following the work of #3275 and #3284

Arsnael commented on src/site/markdown/server/manage-webadmin.md:
```diff
@@ -1151,6 +1152,37 @@ Response codes:
+**WARNING**: this task do not take into account concurrent modifications upon a single current quota recomputation. 
+Rerunning the task will *eventually* provide the consistent result. As such we advise to run this task offline. 
+
+In order to ensure being offline, stop the traffic on SMTP, JMAP and IMAP ports, for example via re-configuration or 
```
From Benoit in #3284 :

> We don't alter source of truth we don't need to be offline.

chibenwa commented on mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/task/RecomputeCurrentQuotasTask.java:
```diff
@@ -0,0 +1,97 @@
+            snapshot.getProcessedQuotaRootCount(),
+            snapshot.getFailedQuotaRoots()
+                .stream()
+                .map(QuotaRoot::toString)
```
We should avoid relying on toString. Can we have a asString method?

chibenwa commented on server/container/guice/protocols/webadmin-mailbox/src/main/java/org/apache/james/modules/server/MailboxRoutesModule.java:
```diff
@@ -58,5 +59,9 @@ protected void configure() {
         Multibinder.newSetBinder(binder(), TaskRegistration.class, Names.named(ONE_MAILBOX_TASKS));
         Multibinder.newSetBinder(binder(), TaskRegistration.class, Names.named(ONE_MAIL_TASKS));
+
+        Multibinder.newSetBinder(binder(), TaskRegistration.class, Names.named(USER_QUOTAS_OPERATIONS_INJECTION_KEY))
```
Memory in theory won't support current quota recomputation. Thus we need the taskToRequest binding to be done in a separate module, being binded only in products that supports it.

Arsnael commented:
> Please contribute webadmin documentation

It is already contributed???

chibenwa commented on src/site/markdown/server/manage-webadmin.md:
```diff
@@ -1151,6 +1152,34 @@ Response codes:
+ 
+### Recomputing current quotas for users
+
+This task is available on top of Guice products.
```
WXebadmin is only available on top of guice product, this totology can be removed.

chibenwa replied:
Not available for memory product I bet

mbaechler commented on server/protocols/webadmin/webadmin-mailbox/pom.xml:
```diff
@@ -68,6 +68,10 @@
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>apache-james-mailbox-store</artifactId>
```
Looks very suspicious to me.

chibenwa replied:
@Arsnael plans to do a `mailbox-tools-quota` maven package that would replace this awefull mailbox-store dependency.

This is a proposal that I personally support.

Arsnael replied:
Done

chibenwa commented:
> Could you tell me what would be the use case for others mailbox implementations?

Unreliable InVMEventBus in the face is hard shut downed.

At least once event delivery messing things up. Current quota is not indempotent.

Failing JPA DB writes out of the mailboxManager transactions.

The failure cases are many as you often says ;-)

Why would we prevent other backends to benefit from @Arsnael nice work? Why would we develop 10 times the same thing?

chibenwa commented:
`MemorySpamAssassinContractTest.expungingSpamMessageShouldNotImpactSpamAssassinState » ConditionTimeout`

test this please

Arsnael commented:
> The JIRA issue doesn't have a definition of done so it's hard to judge if this PR is reaching its goal.

Sorry about that. Just updated it with a DoD and also the link to the proper ADR regarding this task

Arsnael commented:
> However, I doubt we need to recompute mailbox quotas for any other mailbox backend other than Cassandra.
> Could you tell me what would be the use case for others mailbox implementations?

For the reasons addressed by @chibenwa above, and the fact that the corresponding ADR (https://github.com/apache/james-project/blob/master/src/adr/0028-Recompute-mailbox-quotas.md), that we all approved and agreed to, is stating this as well:

```
This way, non-Cassandra implementation will also benefit from this task.
```

chibenwa commented on mailbox/jpa/src/test/java/org/apache/james/mailbox/jpa/mail/task/JPARecomputeCurrentQuotasServiceTest.java:
```diff
@@ -0,0 +1,127 @@
+
+import com.google.common.collect.ImmutableList;
+
+@Disabled("JAMES-3138 Does not work, if someone wants to help...")
```
Let's not add it in the commit then, and let's not expose the task over webadmin for JPA.

Let's wait a kind contributor to fix it.

chibenwa commented on mailbox/spring/src/main/resources/META-INF/spring/spring-mailbox.xml:
```diff
@@ -39,7 +39,7 @@
       Mailbox Copier
      -->
-    <bean id="mailboxcopier" class="org.apache.james.mailbox.tools.copier.MailboxCopierImpl"/>
+    <bean id="mailboxcopier" class="org.apache.james.mailbox.tools.quotas.MailboxCopierImpl"/>
```
oups?

chibenwa commented on mailbox/store/pom.xml:
```diff
@@ -62,11 +62,25 @@
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>james-server-task-json</artifactId>
+        </dependency>
```
Why these dependency changes? Still needed?

Arsnael replied:
I need to recheck you are right

chibenwa commented on mailbox/store/src/main/java/org/apache/james/mailbox/store/quota/CurrentQuotaCalculator.java:
```diff
@@ -26,8 +26,11 @@
 
```
This class deserve to be in tools/quotas ?

chibenwa commented on mailbox/store/src/main/java/org/apache/james/mailbox/store/quota/ListeningCurrentQuotaUpdater.java:
```diff
@@ -19,20 +19,25 @@
 package org.apache.james.mailbox.store.quota;
```
This class deserve to be in tools/quotas ?

chibenwa commented on mailbox/store/src/main/java/org/apache/james/mailbox/store/quota/StoreCurrentQuotaManager.java:
```diff
@@ -20,13 +20,26 @@
 package org.apache.james.mailbox.store.quota;
```
This class deserve to be in tools/quotas ?

chibenwa commented on mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageIdManagerQuotaTest.java:
```diff
@@ -37,6 +37,7 @@
 import org.apache.james.mailbox.quota.CurrentQuotaManager;
```
This class deserve to be in tools/quotas ?

Arsnael replied:
Not sure about this one?

Arsnael replied:
Not sure about this one?

chibenwa commented on mailbox/store/src/test/java/org/apache/james/mailbox/store/quota/ListeningCurrentQuotaUpdaterTest.java:
```diff
@@ -45,6 +45,7 @@
 import org.apache.james.mailbox.events.MailboxListener;
```
This class deserve to be in tools/quotas ?

chibenwa commented on mailbox/store/src/test/java/org/apache/james/mailbox/store/quota/StoreCurrentQuotaManagerTest.java:
```diff
@@ -20,18 +20,24 @@
 package org.apache.james.mailbox.store.quota;
```
This class deserve to be in tools/quotas ?

chibenwa commented on mailbox/tools/copier/src/main/java/org/apache/james/mailbox/tools/quotas/MailboxCopierImpl.java:
```diff
@@ -16,7 +16,7 @@
  * under the License.                                           *
  ****************************************************************/
-package org.apache.james.mailbox.tools.copier;
+package org.apache.james.mailbox.tools.quotas;
```
OUPS?

chibenwa commented on mailbox/tools/copier/src/test/java/org/apache/james/mailbox/tools/quotas/MailboxCopierTest.java:
```diff
@@ -16,7 +16,7 @@
  * under the License.                                           *
  ****************************************************************/
-package org.apache.james.mailbox.tools.copier;
+package org.apache.james.mailbox.tools.quotas;
```
OUPS?

chibenwa commented on mailbox/tools/quotas/pom.xml:
```diff
@@ -0,0 +1,76 @@
+        </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>apache-james-mailbox-store</artifactId>
```
That's evil. Moving all the store related quota class in here should solve this.

Then we could easily makes store depends on tools/quota

Arsnael replied:
Ok problem is, among the classes you asked me to move there... :
* CurrentQuotaCalculator: using `MailboxSessionMapperFactory`, from mailbox-store
* AbstractMessageIdManagerQuotaTest: using `MessageIdManagerTestSystem` from mailbox-store

And if we don't move those... well I still need `CurrentQuotaCalculator` in the service. So hard to see how I can remove that dependency without an other yet refactoring, and tbh tools-indexer is having evil too (didn't say was right though :) )

chibenwa commented on server/container/spring/src/main/resources/META-INF/org/apache/james/spring-server.xml:
```diff
@@ -152,7 +152,7 @@
 
     <!-- Mailbox Copier -->
-    <bean id="mailboxcopier" class="org.apache.james.mailbox.tools.copier.MailboxCopierImpl"/>
+    <bean id="mailboxcopier" class="org.apache.james.mailbox.tools.quotas.MailboxCopierImpl"/>
```
OUPS.

I believe that the copier have nothing to do in quotas package.

Arsnael replied:
Sorry not been careful with Intellij obviously...

chibenwa commented:
Still a bit of mvn module management wizardry to undergo ;-)

mbaechler commented:
> > However, I doubt we need to recompute mailbox quotas for any other mailbox backend other than Cassandra.
> > Could you tell me what would be the use case for others mailbox implementations?
> 
> For the reasons addressed by @chibenwa above, and the fact that the corresponding ADR (https://github.com/apache/james-project/blob/master/src/adr/0028-Recompute-mailbox-quotas.md), that we all approved and agreed to, is stating this as well:
> 
> ```
> This way, non-Cassandra implementation will also benefit from this task.
> ```

Thank you for the links and explanations.

nvduc91 commented on mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraCurrentQuotaManager.java:
```diff
@@ -47,58 +51,70 @@
-            .from(CassandraCurrentQuota.TABLE_NAME)
+        this.getCurrentStorageStatement = session.prepare(select(STORAGE)
+            .from(TABLE_NAME)
+            .where(eq(CassandraCurrentQuota.QUOTA_ROOT, bindMarker())));
```
Small thing but `CassandraCurrentQuota.QUOTA_ROOT` can import static

Arsnael replied:
I think I had it before, seems got lost in the multiples rebases... Thanks for spotting it !

Arsnael commented:
Rebasing on #3284 (again). Reworked the PR to add the wrapping task in the new `apache-james-mailbox-tools-quota-recompute` that we agreed on yesterday, introduced by #3284

mbaechler commented on mailbox/tools/quota-recompute/pom.xml:
```diff
@@ -36,15 +36,35 @@
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>james-server-task-json</artifactId>
```
importing json into mailbox-api?

Arsnael replied:
That's not `mailbox-api` but `mailbox-tools-quota-recompute` module here?
Well to create the task `RecomputeCurrentQuotasTask` wrapping around the service `RecomputeCurrentQuotasService`, with its DTOs that need `james-server-task-json`? `RecomputeCurrentQuotasTaskDTO` and `RecomputeCurrentQuotasTaskAdditionalInformationDTO`)

Can you please then be more explicit in why you think it's wrong again, and what are your suggestions around that? Thanks.

mbaechler replied:
I checked twice before adding this comment: github displayed mailbox-api ... now it's no longer the case.

Arsnael replied:
There is some serious issues with github seems these days...

Arsnael commented:
```
[7f22e8dc55b093110c5264175aa90d2ace930274] [ERROR] Errors: 
[7f22e8dc55b093110c5264175aa90d2ace930274] [ERROR]   CassandraDumbBlobStoreCacheTest.shouldReturnDataWhenCacheSmallDataInConfigurationTTL » ConditionTimeout
```
hmmmmm... test this please

Arsnael commented:
Rebased with master

Arsnael commented on mailbox/api/src/main/java/org/apache/james/mailbox/model/QuotaRoot.java:
```diff
@@ -73,4 +73,9 @@ public String toString() {
     }
+
+    public String asString() {
+        return domain.map(domainValue -> value + "@" + domainValue.asString())
```
Not sure to get your point. I think as I understood QuotaRoot, the `value` is pretty much the composition of:
```
USER_NAMESPACE (like "#private" for example) + SEPARATOR ("#") + username
```
So for the user bob I guess the value would be: `#private&bob`

Then you have the optional `domain` (like james.org or whatever)

The username seems already in value, so I don't get your point here...?

mbaechler replied:
Well, you may be right. Honestly, the only thing I see is `String` so I don't know what's in.
May having a `QuotaRootTest` class that covers `asString` would help?

Arsnael commented:
The tests are green but the build crashed in the dockering phase

test this please

chibenwa commented:
> Do we have integration tests?

We need at least a test showing that the task is well exposed.

Arsnael commented:
> > Do we have integration tests?
> 
> We need at least a test showing that the task is well exposed.

There is one in rabbitmq integration tests (distributed james). But I guess you mean I need for other products as well?

chibenwa commented:
> There is one in rabbitmq integration tests (distributed james). But I guess you mean I need for other products as well?

That's enough for me

mbaechler commented on mailbox/api/src/test/java/org/apache/james/mailbox/model/QuotaRootTest.java:
```diff
@@ -28,4 +33,23 @@
+        Domain domain = Domain.of("apache.org");
+        QuotaRoot quotaRoot = QuotaRoot.quotaRoot(value, Optional.of(domain));
+
+        String expectedValue = value + "@" + domain.asString();
```
please don't use operation to compose expected result ... you should hardcode it in this case

mbaechler commented:
> > > Do we have integration tests?
> > 
> > 
> > We need at least a test showing that the task is well exposed.
> 
> There is one in rabbitmq integration tests (distributed james). But I guess you mean I need for other products as well?

This test computes nothing or I miss something?

chibenwa commented:
> This test computes nothing or I miss something?

We're not doing for other consistency tasks.

Fault injection in integration tests requires more work as they don't rely on testing session.

I propose handling this as a separate PR enhancement.

chibenwa commented:
@mbaechler see https://github.com/linagora/james-project/pull/3329

Arsnael commented:
> I propose handling this as a separate PR enhancement.

cf #3329

mbaechler commented:
> > This test computes nothing or I miss something?
> 
> We're not doing for other consistency tasks.
> 
> Fault injection in integration tests requires more work as they don't rely on testing session.
> 
> I propose handling this as a separate PR enhancement.

I don't request to test inconsistencies but the run on a non-empty server.

chibenwa commented:
> I don't request to test inconsistencies but the run on a non-empty server.

In the end, that's what https://github.com/linagora/james-project/pull/3329 provide (and even more!)

mbaechler commented:
> > I don't request to test inconsistencies but the run on a non-empty server.
> 
> In the end, that's what #3329 provide (and even more!)

Do you have a fix-nothing integration test for non-empty server?

chibenwa commented:
> Do you have a fix-nothing integration test for non-empty server?

For sure, we always can write more tests.

Arsnael commented:
```
[17a9ac75100b94377a12093b60d65ecb1264926a] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.2:test (default-test) on project cassandra-jmap-draft-integration-testing: There was a timeout or other error in the fork -> [Help 1]
```
test this please

Arsnael commented:
> Do you have a fix-nothing integration test for non-empty server?

@mbaechler : This is a remark for #3329 right? If there is no other remarks for this PR, is it possible to approve, or do you have other concerns?

chibenwa commented:
@mbaechler see https://github.com/linagora/james-project/pull/3329/commits/ff0d46c5bf6d68754809609ea30ae913713fb082

test this please

mbaechler commented:
> > Do you have a fix-nothing integration test for non-empty server?
> 
> @mbaechler : This is a remark for #3329 right? If there is no other remarks for this PR, is it possible to approve, or do you have other concerns?

Well, you decided to split things in two PR, that's ok, but please don't blame me if my comments at the origin of this split is in the wrong place.

My point was: I didn't expect the tests @chibenwa wrote, I just wanted an integration test actually demonstrating some things run, in this PR.

> @mbaechler see ff0d46c

Thanks. I'll approve this PR.

chibenwa commented:
Merged
